File: pulsar-skeleton/src/test/java/ai/platon/pulsar/crawl/parse/TestCrawlFilter.java
package ai.platon.pulsar.crawl.parse;

import ai.platon.pulsar.common.Urls;
import ai.platon.pulsar.common.config.MutableConfig;
import ai.platon.pulsar.crawl.filter.CrawlFilters;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import java.io.IOException;
import java.net.MalformedURLException;
import java.util.Map;

import static org.junit.Assert.*;

@Ignore
@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration(locations = {"classpath:/test-context/filter-beans.xml"})
public class TestCrawlFilter {

    private String[] detailUrls = {
            "http://mall.jumei.com/product_200918.html?from=store_lancome_list_items_7_4"
    };

    @Autowired
    MutableConfig conf;
    @Autowired
    private CrawlFilters crawlFilters;

    @Before
    public void setUp() throws IOException {
        String crawlFilterRules = "{}";
        conf.set(CrawlFilters.CRAWL_FILTER_RULES, crawlFilterRules);
    }

    @Test
    public void testKeyRange() throws MalformedURLException {
        Map<String, String> keyRange = crawlFilters.getReversedKeyRanges();

        System.out.println(keyRange);

        assertEquals("com.jumei.mall:http/\uFFFF", keyRange.get("com.jumei.mall:http"));
        assertEquals("com.jumei.lancome:http/search.html\uFFFF", keyRange.get("com.jumei.lancome:http/search.html"));
        assertNotEquals("com.jumei.lancome:http/search.html\\uFFFF", keyRange.get("com.jumei.lancome:http/search.html"));

        for (String detailUrl : detailUrls) {
            assertTrue(crawlFilters.testKeyRangeSatisfied(Urls.reverseUrl(detailUrl)));
        }
    }

    @Test
    public void testMaxKeyRange() throws MalformedURLException {
        String[] keyRange = crawlFilters.getMaxReversedKeyRange();
        System.out.println(keyRange[0] + ", " + keyRange[1]);

        assertEquals(65438, '\uFFFF' - 'a');

        for (String detailUrl : detailUrls) {
            detailUrl = Urls.reverseUrl(detailUrl);

            assertTrue("com.jumei.lancome:http/search.html".compareTo(detailUrl) < 0);
            assertTrue("com.jumei.mall:http/\uFFFF".compareTo(detailUrl) > 0);
            // Note : \uFFFF, not \\uFFFF
            assertFalse("com.jumei.mall:http/\\uFFFF".compareTo(detailUrl) > 0);
        }

        for (String detailUrl : detailUrls) {
            detailUrl = Urls.reverseUrl(detailUrl);
            assertTrue(keyRange[0].compareTo(detailUrl) < 0);
            assertTrue(keyRange[1].compareTo(detailUrl) > 0);
        }
    }
}

File: pulsar-skeleton/src/test/kotlin/ai/platon/pulsar/crawl/parse/TestCrawlFilter.kt
package ai.platon.pulsar.crawl.parse

import ai.platon.pulsar.common.config.MutableConfig
import ai.platon.pulsar.common.urls.UrlUtils
import ai.platon.pulsar.common.urls.UrlUtils.reverseUrl
import ai.platon.pulsar.crawl.filter.CrawlFilters
import kotlin.test.*
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit4.SpringRunner
import java.io.IOException
import java.net.MalformedURLException
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

@Ignore
@RunWith(SpringRunner::class)
@ContextConfiguration(locations = ["classpath:/test-context/filter-beans.xml"])
class TestCrawlFilter {
    private val detailUrls = arrayOf(
            "http://mall.jumei.com/product_200918.html?from=store_lancome_list_items_7_4"
    )

    @Autowired
    lateinit var conf: MutableConfig
    @Autowired
    lateinit var crawlFilters: CrawlFilters

    @BeforeTest
    @Throws(IOException::class)
    fun setUp() {
        val crawlFilterRules = "{}"
        conf[CrawlFilters.CRAWL_FILTER_RULES] = crawlFilterRules
    }

    @Test
    @Throws(MalformedURLException::class)
    fun testKeyRange() {
        val keyRange = crawlFilters.reversedKeyRanges
        println(keyRange)
        assertEquals("com.jumei.mall:http/\uFFFF", keyRange["com.jumei.mall:http"])
        assertEquals("com.jumei.lancome:http/search.html\uFFFF", keyRange["com.jumei.lancome:http/search.html"])
        assertNotEquals("com.jumei.lancome:http/search.html\\uFFFF", keyRange["com.jumei.lancome:http/search.html"])
        for (detailUrl in detailUrls) {
            assertTrue(crawlFilters.testKeyRangeSatisfied(reverseUrl(detailUrl)))
        }
    }

    @Test
    @Throws(MalformedURLException::class)
    fun testMaxKeyRange() {
        val keyRange = arrayOf("\u0000", "\uFFFF")
        println(keyRange[0].toString() + ", " + keyRange[1])
        assertEquals(65438, '\uFFFF' - 'a')
        for (url in detailUrls) {
            val reversedUrl = UrlUtils.reverseUrl(url)
            assertTrue("com.jumei.lancome:http/search.html" < reversedUrl)
            assertTrue("com.jumei.mall:http/\uFFFF" > reversedUrl)
            // Note : \uFFFF, not \\uFFFF
            assertFalse("com.jumei.mall:http/\\uFFFF" > reversedUrl)
        }

        for (url in detailUrls) {
            val reversedUrl = UrlUtils.reverseUrl(url)
            println(reversedUrl)
            assertTrue(keyRange[0] < reversedUrl)
            assertTrue(keyRange[1] > reversedUrl)
        }
    }
}
